### ISSUE TITLE
Door State inverted

### ISSUE THREAD

**Issue by pavandave** [SpanPanel/Span#11]
What happened/changed?

My door sensor state seems to be inverted using this integration. I had someone from the installer come by and configure it so that SPAN could see my Powerwall 3 since that was released after my install and required someone present to configure using the installer apps. After that change was made, a few hours later I noticed that HA showed the door in a Door State of "Tampering Detected". I opened and closed a few times and noticed that it was inverted.  Reloaded the Span in HA and then restarted Home Assistant to no avail. Then I checked the SPAN app and noticed it was actually correct. 

I reset and fully rebooted home assistant and also tried to manually clear the sensor but it immediately flipped back. 

Is there a way to reset this or is this a potential defect? Only thing I haven't done is reset the SPAN but not quite certain on those instructions especially since SPAN app is a accurate.

Versions:
Core 2024.8.2
Supervisor 2024.08.0
Operating System 13.1
Frontend 20240809.0

SPAN software version  - spanos2/r202429/04 - Same version as before the tech came out.
HACS SpanPanel version - 0.0.11

![image](https://github.com/user-attachments/assets/6d4ec5d6-67c0-4977-a5a3-d16718ca9615)
![image](https://github.com/user-attachments/assets/2256ca3e-694d-46fc-806d-2514be440c41)

Thanks,

**Comment by cayossarian** [comment]
@pavandave yes, thank you for reporting that.   Update to the new release 0.0.12

**Comment by bryanembley** [comment]
I'm having this exact issue in 1.0.2 and 2025.1.0. Pavandave documented it perfectly and I reloaded config, restarted HA, removed SPAN integration, restarted HA, added SPAN back, restarted HA, and the issue persists.

**Comment by cayossarian** [comment]
What version of software is your panel?   There was a defect in the firmware that put the door in the wrong state.   We fixed the t n our integration and the SPAN reversed and we did too.  My guess s you have an old firmware.  Mine looks good at spanos2 r202446/03

**Comment by bryanembley** [comment]
spanos2/r202446/03

:/

**Comment by cayossarian** [comment]
I can't reproduce the problem and no one else reported it.  Assume you confirmed the integration version.  Have you tried to reboot Home Assistant?

**Comment by bryanembley** [comment]
Yep. Integration is 1.0.2 and I even wiped it clean off, rebooted, and re-added it thinking that perhaps the three door openings to authorize access in the first place might get the toggle in the right spot. 

This isn't a dealbreaker for me. I'll probably just write a little sensor flip script in my config file for now and then comment it out when it starts to behave. It seems like if I'm the only one with the issue that's probably the way to go.

**Comment by pavandave** [comment]
I have 1.0.2 and spanos2/r202446/03 and do not see this issue.

Will monitor.

**Comment by cayossarian** [comment]
I'll go through the code.  Maybe there is a race condition.

**Comment by cayossarian** [comment]
Would be a big help if you can try the beta where i tried to address any race conditions that may have inverted the door state for @bryanembley.  Got to HACs where your current SPAN integration is installed and select 'Redownload" and it will ask if you want a different version.  Select the latest beta and it will update.  Let me know if the door state situation improves without using additional templates.  I fixed a couple of other minor things like deprecation warning and a config flow initial issue for an invalid token.  Let me know...thanks!

**Comment by bryanembley** [comment]
=( 1.0.3b4 still has the inverted door state for me.

**Comment by cayossarian** [comment]
Can you please leave a screenshot of the sensor for the door state from the SPAN device page?   I want to see the type of sensor it lists as the sensor should be tamper type with clear and detected states, not a door open/close.

**Comment by bryanembley** [comment]
![Screenshot_20250105-103524](https://github.com/user-attachments/assets/6afc49fa-f943-4a61-a4b5-fe44c9795553)
 Like this?

**Comment by cayossarian** [comment]
You can also call the API directly using a REST tool if you know how.  
https://gist.github.com/hyun007/c689fbed10424b558f140c54851659e3

I can see from your screenshot the door is a tamper type and I assume it says detected, not clear, where it lists all the sensors on the device page?

**Comment by bryanembley** [comment]
I do not know how to do the REST thing with the API, but yes, it reports "tampering detected"

**Comment by cayossarian** [comment]
i was looking for the door state from the device page like this:

![top_door](https://github.com/user-attachments/assets/487483f4-6646-4110-8de0-a805a5307828)
![door_state](https://github.com/user-attachments/assets/73ade0bc-3dcc-46b2-9d2b-cddbf358091e)

**Comment by bryanembley** [comment]
![Screenshot_20250105-105407](https://github.com/user-attachments/assets/74cc4887-5049-45a9-9799-7e9a3aee11ea)

**Comment by cayossarian** [comment]
Yeah can you try to call the API directly using that site as a reference?   If you need help let me know

**Comment by bryanembley** [comment]
I need help. I understand that I'm going to manually query the SPAN API and I think I understand that the command you want me to run is 'GET /api/v1/status' but what I don't understand is how to direct that to the place it needs to go. Am I using something within HA? Just a regular command line?

**Comment by bryanembley** [comment]
I kept reading and tried curl http://<panelip>/api/v1/status and got this:

![SpanAPICall](https://github.com/user-attachments/assets/799b8183-d403-4f8c-b2da-e0d90be7f1cd)

Is that what you needed?

**Comment by cayossarian** [comment]
Thats's interesting  because I get this: 

{"software":{"firmwareVersion":"spanos2/r202446/03","updateStatus":"idle","env":"prod"},"system":{"manufacturer":"Span","serial":"nj-xxxx-xxxx","model":"00200","doorState":"UNKNOWN","proximityProven":false,"uptime":1707154},"network":{"eth0Link":true,"wlanLink":true,"wwanLink":true}}%       


@pavandave, what do you see?

**Comment by bryanembley** [comment]
Well...data helps drive a solution. Does it show "Unknown" for the door when you point a browser to your panel's local IP? For mine, all that same information is on a web page with a "Emergency Reconnect" button at the bottom but the door state says "Closed"
![spanweb](https://github.com/user-attachments/assets/5ec8de0f-642e-4865-985e-f94e233bb984)

**Comment by cayossarian** [comment]
So i opened and closed the door and without going to the app in the browser checked with CURL again and the door then showed closed and the integration showed tampering.  

So what happened is that the panel API is losing its door state if the door is not operated for a time or some other condition causes it to go brain dead.  When this issue originally occurred we misdiagnosed  the issue and changed the behavior of our integration in error.  

Appreciate you helping run that down.  I'll be another fix in today and publish a beta if you can test that when I notify you.  If the door state is UNKNOWN we will show the sensor accordingly unavailable and report the correct state in either open or closed.

**Comment by bryanembley** [comment]
I'm glad I could help, meager and halting as my contribution was.

**Comment by cayossarian** [comment]
OK beta 5 is posted.  give it a spin if you don't mind, thanks

**Comment by bryanembley** [comment]
Bingo! Now I have to go wipe out the value switch script in my configuration.yaml 

Thanks!

**Comment by cayossarian** [comment]
@pavandave your data is probably also coming back as unknown so without touching the door it would be great if you tried beta 5.  The sensor should then reflect an unknown value until you open the door

**Comment by pavandave** [comment]
Installed, showed as tamper detected on HA restart. I opened and closed
panel and it cleared.

**Comment by cayossarian** [comment]
Thanks for checking @pavandave, I have adjusted the final release so that the sensor is unavailable if the panel returns UNKNOWN.  Unfortunately this is a SPAN defect so nothing we can do about it in that case.   The door state properly set if the returned panel value is open or closed.   Fixed in 1.0.3
